### PR TITLE
HAWQ-1409. Send AGG-TYPE header to PXF

### PR DESCRIPTION
**Pull request by kavinderd**
This change is mean to be a proof of concept that pushing down
aggregate function information from HAWQ to the underlying external
storage layer does indeed improve performance

### CODE REVIEW

File: src/backend/access/external/pxfheaders.c [hunk 1]
```diff
@@ -110,6 +110,20 @@ void build_http_header(PxfInputData *input)
 	else
 		churl_headers_append(headers, "X-GP-HAS-FILTER", "0");
 
+	/* Aggregate information */
+	if (input->agg_type) {
+		char agg_groups_str[sizeof(int32)];
```
Comment (hsyuan) [inline src/backend/access/external/pxfheaders.c]: Where is this var used?

File: src/backend/executor/nodeAgg.c [hunk 2]
```diff
@@ -1950,6 +1950,21 @@ ExecInitAgg(Agg *node, EState *estate, int eflags)
...
+	if (outerPlan->type == T_ExternalScan) {
+		/*
+		 * Hack to indicate to PXF when there is an external scan
+		 */
+		if (aggstate->aggs && list_length(aggstate->aggs) == 1) {
+			foreach(l, aggstate->aggs) {
+				AggrefExprState *aggrefstate = (AggrefExprState *) lfirst(l);
+				Aggref	   *aggref = (Aggref *) aggrefstate->xprstate.expr;
+				//Only dealing with one agg
+				if (aggref->aggfnoid == 2147 || aggref->aggfnoid == 2803) {
```
Comment (hsyuan) [inline src/backend/executor/nodeAgg.c]: Would be better use `COUNT_ANY_OID` and `COUNT_STAR_OID`...

File: src/backend/access/external/fileam.c [hunk 3]
```diff
@@ -459,12 +459,20 @@ external_stopscan(FileScanDesc scan)
...
+	Plan *rootPlan;
 
 	if (state != NULL)
 	{
 		desc->projInfo = state->ps_ProjInfo;
+		/*
+		 * If we have an agg type then our parent is an Agg node
+		 */
+		rootPlan = state->state->es_plannedstmt->planTree;
+		if (rootPlan->type == T_Agg && es_state->parent_agg_type) {
```
Comment (hsyuan) [inline src/backend/access/external/fileam.c]: Recommend to use `IsA(rootPlan, Agg)`

File: src/backend/executor/nodeAgg.c [hunk 4]
```diff
@@ -1950,6 +1950,21 @@ ExecInitAgg(Agg *node, EState *estate, int eflags)
 	 * initialize child nodes
 	 */
 	outerPlan = outerPlan(node);
+	if (outerPlan->type == T_ExternalScan) {
```
Comment (hsyuan) [inline src/backend/executor/nodeAgg.c]: Better use `IsA(outerPlan, ExternalScan)`

File: src/backend/executor/nodeAgg.c [hunk 5]
```diff
@@ -1950,6 +1950,21 @@ ExecInitAgg(Agg *node, EState *estate, int eflags)
...
+		/*
+		 * Hack to indicate to PXF when there is an external scan
+		 */
+		if (aggstate->aggs && list_length(aggstate->aggs) == 1) {
```
Comment (hsyuan) [inline src/backend/executor/nodeAgg.c]: Do we need check `aggstate->aggs` is null or not? if it is null, `list_length` will return 0.

File: src/backend/executor/nodeAgg.c [hunk 6]
```diff
@@ -1950,6 +1950,21 @@ ExecInitAgg(Agg *node, EState *estate, int eflags)
...
+			foreach(l, aggstate->aggs) {
```
Comment (hsyuan) [inline src/backend/executor/nodeAgg.c]: if the size of `aggstate->aggs` is already 1, do we need use `foreach` to iterate?

File: src/backend/executor/nodeAgg.c [hunk 7]
```diff
@@ -1950,6 +1950,21 @@ ExecInitAgg(Agg *node, EState *estate, int eflags)
...
+				AggrefExprState *aggrefstate = (AggrefExprState *) lfirst(l);
```
Comment (hsyuan) [inline src/backend/executor/nodeAgg.c]: just by changing `lfirst(l)` to `linitial(aggstate->aggs)`, you can remove `foreach` loop.

File: src/backend/executor/nodeAgg.c (hunk 5 above)
Comment (kavinderd) [inline src/backend/executor/nodeAgg.c]: Oh ok, I thought `list_length()` on a null value would cause an error

**Pull request closed without merging**